### ISSUE TITLE
how to implement background location(when device terminated)?

### ISSUE THREAD

**Issue by guvanch07** [Baseflow/flutter-geolocator#1212]
## 💬 Questions and Help
how to implement background location(when device terminated)?

I see a lot of packages for background location but is it impossible to implement this just using geolocator package?

**Comment by tulioccalazans** [comment]
What do you mean by "when the device turns off"?

**Comment by guvanch07** [comment]
oh I wrote it wrong but I edited, I mean 'when app terminated'

**Comment by tulioccalazans** [comment]
It is possible to get background location even when the app is terminated.
Just follow the steps described in: [geolocator 9.0.2](https://pub.dev/packages/geolocator#listen-to-location-updates)

The important part to make this possible is to set the foreground notification.

If you start the service (Geolocator.getPositionStream) before closing the app, this foreground notification will keep the service/app alive.


```
   foregroundNotificationConfig: const ForegroundNotificationConfig(
        notificationText:
        "Example app will continue to receive your location even when you aren't using it",
        notificationTitle: "Running in Background",
        enableWakeLock: true,
   )
```

**Comment by guvanch07** [comment]
yes I am using Geolocator.getPositionStream and setted locationSettings like you show me below, also notification icon appear but when I terminated the app, stop to come location data(lat,lon and etc) to my database
```
Geolocator.getPositionStream(locationSettings: gmsLocationSettings())
        .listen((position) {
      log('⚙️ \x1B[35m-------> init');
      log('${position.latitude}');
      log('${position.longitude}');
      log('${position.heading}');
     
      final data = LocationData(
        lat: position.latitude,
        lng: position.longitude,
      );
      apiCall(data); // here post request
    });
```

**Comment by tulioccalazans** [comment]
You're right, I just tested the coordinate capture when the app is terminated and although the foreground service notification remains active, the coordinate capture stops working.

I'm looking into this further, but a response from the developers would be nice.

**Comment by guvanch07** [comment]
Always When requesting this permission the user needs to grant permission for the locationWhenInUse permission first, clicking on the Àllow While Using App option on the popup. After allowing the permission the user can request the locationAlways permission and can click on the Change To Always Allow option.

I also request always permission

```
Future<bool> requestAlwaysPermission() async {
    try {
      final status = await Permission.locationAlways.request();
      if (status.isGranted) {
        return true;
      } else {
        return false;
      }
    } catch (e) {
      log(e.toString());
      return false;
    }
  }
```
I call below method after calling await geolocatorPlatform.requestPermission();

**Comment by tulioccalazans** [comment]
Thanks,

I just did it! But now, when I kill the app, I got the message on logcat:

> 2023-02-10 16:31:01.140  7341-7341  FlutterJNI              com.test.app       W  Tried to send a platform message to Flutter, but FlutterJNI was detached from native C++. Could not send. Channel: flutter.baseflow.com/geolocator_updates_android. Response ID: 99

So, When I try to open the app, it freezes with the following message:

> 2023-02-10 16:31:03.576  7341-7341  FlutterGeolocator       com.test.app       D  Flutter engine disconnected. Connected engine count 1
> 2023-02-10 16:31:03.578  7341-7341  FlutterGeolocator       com.test.app       D  Disposing Geolocator services
> 2023-02-10 16:31:03.578  7341-7341  FlutterGeolocator       com.test.app       E  Geolocator position updates stopped
> 2023-02-10 16:31:03.578  7341-7341  FlutterGeolocator       com.test.app       E  There is still another flutter engine connected, not stopping location service

**Comment by tulioccalazans** [comment]
@guvanch07 

I get the following messages when I kill the app:

> D/FlutterGeolocator(28354): Detaching Geolocator from activity
> E/FlutterGeolocator(28354): Geolocator position updates stopped
> E/FlutterGeolocator(28354): There is still another flutter engine connected, not stopping location service
> D/TSBackgroundFetch(28354): ☯️  HeadlessMode? true
> D/FlutterGeolocator(28354): Flutter engine disconnected. Connected engine count 2
> D/FlutterGeolocator(28354): Disposing Geolocator services
> D/FlutterGeolocator(28354): Tried to stop listening when no MethodChannel had been initialized.

As I said, the foreground service notification continues and apparently the execution too, because of this message:

> W/FlutterJNI(32329): Tried to send a platform message to Flutter, but FlutterJNI was detached from native C++. Could not send. Channel: flutter.baseflow.com/geolocator_updates_android. Response ID: 77

Geolocator is not prepared to receive locations when the app is terminated.

I checked the class "flutter-geolocator\geolocator_android\android\src\main\java\com\baseflow\geolocator\geolocator_platform_interface.kt" and noticed that the location service is being disposed when the FlutterEngine is terminated (which happens when kill the app).

```
  @Override
  public void onDetachedFromEngine(@NonNull FlutterPluginBinding binding) {
    unbindForegroundService(binding.getApplicationContext());
    dispose();
  }
```

This might work if we can somehow avoid disposing of the location mechanism. Or we somehow share the flutterEngine with the foregroundService. But I still have no idea how to do that.

Unfortunately I'm thinking of an alternative plugin to replace the geolocator. If you find any, please let me know.

**Comment by guvanch07** [comment]
when I kill the app, connection via usb didn't disconnect with device
continue listen but there are not geo data


V/AudioManager(27994): querySoundEffectsEnabled...
W/libEGL  (27994): EGLNativeWindowType 0x7cc87f9010 disconnect failed
W/FlutterJNI(27994): Tried to send a platform message to Flutter, but FlutterJNI was detached from native C++. Could not send. Channel: flutter.baseflow.com/geolocator_updates_android. Response ID: 554
W/FlutterJNI(27994): Tried to send a platform message to Flutter, but FlutterJNI was detached from native C++. Could not send. Channel: flutter.baseflow.com/geolocator_updates_android. Response ID: 555
W/FlutterJNI(27994): Tried to send a platform message to Flutter, but FlutterJNI was detached from native C++. Could not send. Channel: flutter.baseflow.com/geolocator_updates_android. Response ID: 556
W/FlutterJNI(27994): Tried to send a platform message to Flutter, but FlutterJNI was detached from native C++. Could not send. Channel: flutter.baseflow.com/geolocator_updates_android. Response ID: 557
W/FlutterJNI(27994): Tried to send a platform message to Flutter, but FlutterJNI was detached from native C++. Could not send. Channel: flutter.baseflow.com/geolocator_updates_android. Response ID: 558
W/FlutterJNI(27994): Tried to send a platform message to Flutter, but FlutterJNI was detached from native C++. Could not send. Channel: yandex_mapkit/yandex_map_0. Response ID: 559
W/FlutterJNI(27994): Tried to send a platform message to Flutter, but FlutterJNI was detached from native C++. Could not send. Channel: flutter.baseflow.com/geolocator_updates_android. Response ID: 560
W/FlutterJNI(27994): Tried to send a platform message to Flutter, but FlutterJNI was detached from native C++. Could not send. Channel: flutter.baseflow.com/geolocator_updates_android. Response ID: 561
W/FlutterJNI(27994): Tried to send a platform message to Flutter, but FlutterJNI was detached from native C++. Could not send. Channel: flutter.baseflow.com/geolocator_updates_android. Response ID: 562
W/FlutterJNI(27994): Tried to send a platform message to Flutter, but FlutterJNI was detached from native C++. Could not send. Channel: flutter.baseflow.com/geolocator_updates_android. Response ID: 563
W/FlutterJNI(27994): Tried to send a platform message to Flutter, but FlutterJNI was detached from native C++. Could not send. Channel: flutter.baseflow.com/geolocator_updates_android. Response ID: 564
W/FlutterJNI(27994): Tried to send a platform message to Flutter, but FlutterJNI was detached from native C++. Could not send. Channel: flutter.baseflow.com/geolocator_updates_android. Response ID: 565
W/FlutterJNI(27994): Tried to send a platform message to Flutter, but FlutterJNI was detached from native C++. Could not send. Channel: flutter.baseflow.com/geolocator_updates_android. Response ID: 566
W/FlutterJNI(27994): Tried to send a platform message to Flutter, but FlutterJNI was detached from native C++. Could not send. Channel: flutter.baseflow.com/geolocator_updates_android. Response ID: 567

**Comment by tulioccalazans** [comment]
@guvanch07 

I decided to use this lib [background_locator_2](https://pub.dev/packages/background_locator_2), but I needed to make some adjustments because it doesn't work with flutter 3.7.0, so take a look at my fork [FORK](https://github.com/tulioccalazans/background_locator_fixed) until they merge my PR.

You can use the library this way:

```
  background_locator_2:
    git:
      url: https://github.com/tulioccalazans/background_locator_fixed.git
      ref: master
```

I think this library is a good reference for the geolocator team to get some ideas for improvements, because it seems that few things remain to be implemented for the geolocator to work when the application is finished.

**Comment by guvanch07** [comment]
@tulioccalazans 
okay, thank you)

**Comment by Alvish0407** [comment]
> 

Hey @tulioccalazans 

does your [FORK](https://github.com/tulioccalazans/background_locator_fixed) work on iOS when app is terminated ? I have been wandering on internet to find the package for iOS which can fetch the location even when app is terminated. But didn't succeed well.

Currently for android i am using [https://github.com/caiobraga/background_locator](https://github.com/caiobraga/background_locator) which is FORK of [https://github.com/rekabhq/background_locator](https://github.com/rekabhq/background_locator) which seems to be working fine. Even though, i will try your package too.

I found [flutter_background_geolocation](https://pub.dev/packages/flutter_background_geolocation), it works fine on simulator but not on real device though. I can say that this library on iOS uses fitness and motion tracking to fetch the location.

Help would be much appreciated.

**Comment by tulioccalazans** [comment]
> > 
> 
> Hey @tulioccalazans
> 
> does your [FORK](https://github.com/tulioccalazans/background_locator_fixed) work on iOS when app is terminated ? I have been wandering on internet to find the package for iOS which can fetch the location even when app is terminated. But didn't succeed well.
> 
> Currently for android i am using https://github.com/caiobraga/background_locator which is FORK of https://github.com/rekabhq/background_locator which seems to be working fine. Even though, i will try your package too.
> 
> I found [flutter_background_geolocation](https://pub.dev/packages/flutter_background_geolocation), it works fine on simulator but not on real device though. I can say that this library on iOS uses fitness and motion tracking to fetch the location.
> 
> Help would be much appreciated.

My FORK works perfectly on Android devices. I'm in the testing phase on iOS devices. But I believe it wouldn't be a problem since the part that needed to be changed was the flutter part and not the native part of the background_locator_2 library. I will carry out the tests in the next few days and any news, I will report here!

**Comment by christocracy** [comment]
> I found flutter_background_geolocation, it works fine on simulator but not on real device though

It works perfectly on real devices for the last 9 years (since iOS 7).  I test it every day.  It never fails.

**Comment by moudmagongore** [comment]
good evening @tulioccalazans  I need your help, in fact when I configure the plugin as you did and I try to take the example of the background_locator_2 documentation, my editor underlines me it is import below

import 'file_manager.dart';
import 'location_callback_handler.dart';
import 'location_service_repository.dart';

**Comment by moudmagongore** [comment]
@tulioccalazans I have installed the latest version of background_locator_2, and I'm trying to use the example from the documentation in a flutter project but I'm still getting errors with my editor pointing out to me it's 3 import below

import 'file_manager.dart';
import 'location_callback_handler.dart';
import 'location_service_repository.dart';

Is it possible to help me how to implement this package in a flutter project ?

**Comment by tulioccalazans** [comment]
> hello @tulioccalazans i need your help, i am using the latest version of backgroun_locator_2, everything works as expected on android emulator, but when i build the app and i install it in android phone, every time i run the application it closes automatically. Can you help me please ?

@moudmagongore Sure, please open an issue in background_locator_2 project giving more details about this error.

Use this link to create it: [background_locator_2  issues](https://github.com/Yukams/background_locator_fixed/issues) 

Attach the isse link in response to this issue so I can find it.

**Comment by moudmagongore** [comment]
@tulioccalazans  here is the link https://github.com/Yukams/background_locator_fixed/issues/84

**Comment by progressify** [comment]
up

**Comment by MSF01** [comment]
Bump... Any progress on this case?

**Comment by Mohamed50** [comment]
up

**Comment by Ilyes-git** [comment]
up
